datafusion/functions/src/string/concat.rs

```diff
@@ -88,37 +145,91 @@ impl ScalarUDFImpl for ConcatFunc {
         &self.signature
     }
 
+    fn coerce_types(&self, arg_types: &[DataType]) -> Result<Vec<DataType>> {
+        use DataType::*;
+
+        if arg_types.is_empty() {
+            return plan_err!("concat requires at least one argument");
+        }
+
+        let has_arrays = arg_types
+            .iter()
+            .any(|dt| matches!(dt, List(_) | LargeList(_) | FixedSizeList(_, _)));
+        let has_non_arrays = arg_types
+            .iter()
+            .any(|dt| !matches!(dt, List(_) | LargeList(_) | FixedSizeList(_, _) | Null));
+
+        if has_arrays && has_non_arrays {
+            return plan_err!(
+                "Cannot mix array and non-array arguments in concat function. \
+                Use concat(array1, array2, ...) for arrays or concat(str1, str2, ...) for strings, but not both."
```

Comment:
```suggestion
                "Cannot mix array and non-array arguments in concat function.
```

Remove redundant information